### PR TITLE
UPnP Port Forwarding

### PR DESCRIPTION
**Pull request by katekerllenevich**
This pull request allows people with supported routers to use UPnP to automatically port forward (part of #478).

I was unable to test this as my router doesn't support UPnP, so if anyone who can use UPnP would test it that would be nice.

### CODE REVIEW

File: patches/server/0232-UPnP-Port-Forwarding.patch [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+     public volatile Thread shutdownThread; // Paper
+     public volatile boolean abnormalExit = false; // Paper
+@@ -1027,6 +1028,15 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
+         // CraftBukkit end
+         MinecraftServer.LOGGER.info("Stopping server");
+         MinecraftTimings.stopServer(); // Paper
++        // Purpur start
++        if (upnp) {
++            if (me.notom3ga.upnp4j.UPnP4J.close(this.getPort(), me.notom3ga.upnp4j.util.Protocol.TCP)) {
++                LOGGER.info("[UPnP] Port " + this.getPort() + " closed");
```
Comment (mikroskeem) [inline patches/server/0232-UPnP-Port-Forwarding.patch]: `LOGGER.info("[UPnP] Port {} closed", this.getPort());`

File: patches/server/0232-UPnP-Port-Forwarding.patch [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
++            } else {
++                LOGGER.error("[UPnP] Failed to close port " + this.getPort());
```
Comment (mikroskeem) [inline patches/server/0232-UPnP-Port-Forwarding.patch]: `LOGGER.error("[UPnP] Failed to close port {}", this.getPort());`

File: patches/server/0232-UPnP-Port-Forwarding.patch [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+index a75d1872d63aeba5376d787336e4793e813ba21f..e1cca7fbf6db8b1abddb491b0099a2277c4c6380 100644
+--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
++++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+@@ -281,6 +281,31 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
+             return false;
+         }
+ 
++        // Purpur start
++        if (net.pl3x.purpur.PurpurConfig.useUPnP) {
++            LOGGER.info("[UPnP] Attempting to start UPnP port forwarding service...");
```
Comment (mikroskeem) [inline patches/server/0232-UPnP-Port-Forwarding.patch]: `LOGGER.info("[UPnP] Starting UPnP port forwarding service");`

**Review by mikroskeem — CHANGES_REQUESTED** [review]
There are more logger nitpicks, ones I pointed out should be good examples to fix remaining.

**Comment by ATMmachin3** [comment]
Question:
Does this utilise the port the server has been set to use or only the default port?

**Comment by katekerllenevich** [comment]
the port the server has been set to you

File: patches/server/0232-UPnP-Port-Forwarding.patch [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
+index e74ddf6b72dbd335f8a06f05341d5a4d99afa428..1d583c0964bd7cdaab081d112336842d7a1c6db4 100644
+--- a/src/main/java/net/minecraft/server/MinecraftServer.java
++++ b/src/main/java/net/minecraft/server/MinecraftServer.java
+@@ -309,6 +309,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
+     public boolean lagging = false; // Purpur
+     public final SlackActivityAccountant slackActivityAccountant = new SlackActivityAccountant();
+     // Spigot end
++    protected boolean upnp = false; // Purpur
```
Comment (Oharass) [inline patches/server/0232-UPnP-Port-Forwarding.patch:15]: comments

**Review by Oharass — CHANGES_REQUESTED** [review]

**Review by Oharass — APPROVED** [review]
no wrong LGTM

**Comment by katekerllenevich** [comment]
If this is “no wrong” you would have checked and know that the domain for the repo isn’t in use anymore. So really there is much wrong lol

**Comment by Oharass** [comment]
This is a big issue. There is no wrongs, but will not work.

**Comment by katekerllenevich** [comment]
Well duh! It’s my code that I wrote, I think I would know. Also this hasn’t been tested so there could very well be “wrongs”

**Comment by Oharass** [comment]
Sorry perhaps I'm not knowing correctly.

**Comment by CyberFlameGO** [comment]
Interesting feature, I've not seen this implemented in a server before

**Comment by katekerllenevich** [comment]
Just waiting for @BillyGalbreath

File: patches/server/0258-UPnP-Port-Forwarding.patch [hunk 5]
```diff
@@ -0,0 +1,82 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Ben Kerllenevich <ben@omega24.dev>
+Date: Sun, 11 Jul 2021 10:22:55 -0400
```
Comment (BillyGalbreath) [inline patches/server/0258-UPnP-Port-Forwarding.patch]: https://github.com/pl3xgaming/Purpur/blob/ver/1.15/patches/server/0088-UPnP-Port-Forwarding-Service.patch#L2-L3 ;)

Comment (katekerllenevich) [reply patches/server/0258-UPnP-Port-Forwarding.patch]: Must have missed that

**Review by encode42 — APPROVED** [review]

**Pull request merged**